### ISSUE TITLE
No PHP code was found warning in PHP file

### ISSUE THREAD

**Issue by paul-gene** [squizlabs/PHP_CodeSniffer#2507]
I'm running the Magento 2 Coding Standards and I get the weak warning "No PHP code was found in this file..." when inside a PHP file.

The file contains a class and has various arrays that contain HTML strings.

I've tried with the string as just a string and as a NOWDOC, but neither fix the problem.

**Comment by jrfnl** [comment]
@paul-gene Without a code sample/link to the problem file, there is not much anyone can with that information...
Also: using which PHPCS version on which PHP version ? (and possibly _on which OS ?_ if that may be relevant)

**Comment by paul-gene** [comment]
@jrfnl well, the fact that the sniff is showing this warning in a PHP file should be enough? That surely shouldn't happen?!

```
$fieldset->addField(
            'exp_container',
            'note',
            [
                'label' => __('Expiration Date'),
                'required' => true,
                'text' => <<<'TEXT'
<div class="hosted-date-wrap">
    <div class="admin__control-text hosted-control hosted-date" id="exp_month"></div>
    <div class="admin__control-text hosted-control hosted-date" id="exp_year"></div>
</div>
TEXT
            ]
        );
```

I'm on Mac OS 10.13. , PHP 7.1.25 via Homebrew (but using PHP 7.0 interpreter in PHPStorm), PHPCS 3.3.1

**Comment by jrfnl** [comment]
@paul-gene Silly question, but does the file contain a PHP open tag ?

**Comment by paul-gene** [comment]
It does, it's a Class file. Starts with <?php and has a Class with a few methods defined. Code I posted is part of a method and is what is flagged.

**Comment by jrfnl** [comment]
@paul-gene That sounds like it's not the standard PHPCS message, but a message coming from a sniff. Could you run phpcs with the `-s` flag to find out what the error code is of the message (and post it here) ?

**Comment by gsherwood** [comment]
The error that PHPCS includes is: `No PHP code was found in this file and short open tags are not allowed by this install of PHP. This file may be using short open tags but PHP does not allow them.`

You say that the file contains a standard `<?php` open tag, but can you confirm it actually does? Can you replicate this error by using an included standard like PSR-2 to rule out any custom standard issues?

**Comment by paul-gene** [comment]
@gsherwood the file definitely had the `<?php` opening tags.

I'll test with different standards ASAP.

**Comment by gsherwood** [comment]
@paul-gene I think there must be something different about our PHP setups given I can't replicate this, or some invisible character around the opening tag that is causing the PHP tokenizer to fail.

Can you please run: phpcs --standard=PSR2 -vvv --no-cache /path/to/file
and paste the output please (it will be pretty big but give me a clear idea of what PHPCS is doing)

If possible, can you also post a link to the file so I can download it directly from the source.

**Comment by paul-gene** [comment]
@gsherwood So I ran PHPCS via terminal with the flags you suggested, and it's only showing 1 error, which is a method name using a single underscore (magento 2, typical)

I forgot to mention, the No Code sniff is showing in PHPStorm, if that helps?!

**Comment by gsherwood** [comment]
> I forgot to mention, the No Code sniff is showing in PHPStorm, if that helps?!

I downloaded the latest PHPStorm and ran the PSR2 standard over your file. I can replicate the error you are seeing, but only through PHPStorm.

I added some debug code and found that PHPStorm's integration is running PHPCS 3 times over that file when I run the inspections over it. PHPStorm obviously copies the file to a temp location then tells PHPCS to run over that file. The first  2 times, the temp file contains the full content. The 3rd time, the temp file contains just the nowdoc content:

```
[22-May-2019 10:18:02 Australia/Sydney] PHPCS BIN CALLED
[22-May-2019 10:18:02 Australia/Sydney] CHECKING LOCAL FILE /private/var/folders/n3/c0vlv7dn5mld6szhvf7zzynr0000gq/T/phpcs_temp.tmp/temp.php
[22-May-2019 10:18:02 Australia/Sydney] <?php
$fieldset->addField(
    'exp_container',
    'note',
    [
        'label' => __('Expiration Date'),
        'required' => true,
        'text' => <<<'TEXT'
<div class="hosted-date-wrap">
    <div class="admin__control-text hosted-control hosted-date" id="exp_month"></div>
    <div class="admin__control-text hosted-control hosted-date" id="exp_year"></div>
</div>
TEXT
    ]
);

[22-May-2019 10:18:02 Australia/Sydney] PHPCS BIN CALLED
[22-May-2019 10:18:03 Australia/Sydney] CHECKING LOCAL FILE /private/var/folders/n3/c0vlv7dn5mld6szhvf7zzynr0000gq/T/phpcs_temp.tmp/temp.php
[22-May-2019 10:18:03 Australia/Sydney] <?php
$fieldset->addField(
    'exp_container',
    'note',
    [
        'label' => __('Expiration Date'),
        'required' => true,
        'text' => <<<'TEXT'
<div class="hosted-date-wrap">
    <div class="admin__control-text hosted-control hosted-date" id="exp_month"></div>
    <div class="admin__control-text hosted-control hosted-date" id="exp_year"></div>
</div>
TEXT
    ]
);

[22-May-2019 10:18:03 Australia/Sydney] PHPCS BIN CALLED
[22-May-2019 10:18:03 Australia/Sydney] CHECKING LOCAL FILE /private/var/folders/n3/c0vlv7dn5mld6szhvf7zzynr0000gq/T/phpcs_temp.tmp/temp.php
[22-May-2019 10:18:03 Australia/Sydney] <div class="hosted-date-wrap">
    <div class="admin__control-text hosted-control hosted-date" id="exp_month"></div>
    <div class="admin__control-text hosted-control hosted-date" id="exp_year"></div>
</div>
```

I don't know if there is a setting somewhere to tell PHPStorm to only run once, or if there is a reason it inspects the nowdoc using PHPCS directly, or a reason why it only uses that 3rd run to produce the final report.

I also tried this with a custom standard and it worked fine, so I'm lost as to what is going on here. It appears that PHPCS itself is working correctly though, so I'm not sure if I'm going to be able to fix this. I'll have a bit more of a play around and see if I can see anything, but it might be something that needs reporting to PHPStorm itself to see if they know what is going on.

**Comment by gsherwood** [comment]
Sorry, I was wrong. It's not working with other standards. It's just showing that the file was checked twice (once with code and once with the no code error shown). I've got no idea what's going on or how to fix it.
![Screen Shot 2019-05-22 at 10 27 14 am](https://user-images.githubusercontent.com/185976/58139054-68599100-7c7c-11e9-8ce5-421f5bf0a68f.png)

**Comment by paul-gene** [comment]
At least I know I'm not going mad :)

**Comment by gsherwood** [comment]
Even this file fails because PHPStorm sends the HTML string to PHPCS directly for the 3rd run:
```
<?php
foo('<div></div>');
```
There must be some setting somewhere to turn this off, but I can't find it.

**Comment by gsherwood** [comment]
Note that this only happens when you run inspections manually (either in batch, or running the PHPCS inspection by name). The inspection that runs when the file is saved works fine as it only sends the entire file contents once.

I think this might be where I run out of ways to help though. I've tried looking for ways to change this behaviour but I can't find it. I don't actually use PHPStorm, so I'm running with a very default config, so assume this should work fine.

I think you'll need to submit this to PHPStorm and see if they can help you resolve this issue.

**Comment by gsherwood** [comment]
I'm closing this as I can't see anything I can do to help fix it. If I'm wrong, please comment here and I'll look into any further suggestions.

**Comment by jovyllebermudez** [comment]
Im having this error too. will try to completely ignore the file.
